File: src/data/courses/dsa/dynamic-programming/classic-dp/unbounded-knapsack/code/python/unbounded_knapsack.py
class Solution:
    def unboundedKnapsack(self, W, wt, val):
        dp = [0] * (W + 1)
        for i in range(len(wt)):
            for w in range(wt[i], W+1):  # Left to right = reuse
                dp[w] = max(dp[w], dp[w-wt[i]] + val[i])
        return dp[W]

File: src/data/courses/dsa/dynamic-programming/classic-dp/unbounded-knapsack/code/cpp/unbounded_knapsack.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    int unboundedKnapsack(int W, vector<int>& wt, vector<int>& val) {
        vector<int> dp(W + 1, 0);
        for (int i = 0; i < (int)wt.size(); i++)
            for (int w = wt[i]; w <= W; w++) // LEFT to RIGHT — allows reuse
                dp[w] = max(dp[w], dp[w - wt[i]] + val[i]);
        return dp[W];
    }
};
